cirq-google/cirq_google/engine/abstract_local_processor.py

```diff
@@ -0,0 +1,410 @@
+        processor_id: Unique string id of the processor.
+        engine: The parent `AbstractEngine` object, if available.
+        expected_down_time: Optional datetime of the next expected downtime.
+            For informational purpose only.yy
+        expected_recovery_time: Optional datetime when the processor is
+            expected to be available again.  For informational purpose only.
+        schedule:  List of time slots that the scheduling/reservation should
+            use.  All time slots must be non-overlapping.
+        project_name: A project_name for resource naming.
+    """
+
+    def __init__(
+        self,
+        *,
+        processor_id: str,
+        engine: Optional['AbstractEngine'] = None,
+        expected_down_time: Optional[datetime.datetime] = None,
+        expected_recovery_time: Optional[datetime.datetime] = None,
+        schedule: Optional[List[qtypes.QuantumTimeSlot]] = None,
+        project_name: str = 'fake_project',
+    ):
+        self._engine = engine
+        self._expected_recovery_time = expected_recovery_time
+        self._expected_down_time = expected_down_time
+        self._reservations: Dict[str, qtypes.QuantumReservation] = {}
+        self._reservation_id_counter = 0
+        self._processor_id = processor_id
+        self._project_name = project_name
+
+        if schedule is None:
+            self._schedule = [
+                qtypes.QuantumTimeSlot(
+                    processor_name=self._processor_id,
+                    slot_type=qenums.QuantumTimeSlot.TimeSlotType.UNALLOCATED,
+                )
+            ]
+        else:
+            self._schedule = copy.copy(schedule)
+            self._schedule.sort(key=lambda t: t.start_time.seconds or -1)
+
+        for idx in range(len(self._schedule) - 1):
+            if self._schedule[idx].end_time.seconds > self._schedule[idx + 1].start_time.seconds:
+                raise ValueError('Time slots cannot overlap!')
+
+    @property
+    def processor_id(self) -> str:
+        """Unique string id of the processor."""
+        return self._processor_id
+
+    def engine(self) -> Optional['AbstractEngine']:
+        """Returns the parent Engine object.
+
+        Returns:
+            The program's parent Engine.
+
+        Raises:
+            ValueError: if no engine has been defined for this processor.
+        """
+        return self._engine
+
+    def set_engine(self, engine):
+        """Sets the parent processor."""
+        self._engine = engine
+
+    def expected_down_time(self) -> 'Optional[datetime.datetime]':
+        """Returns the start of the next expected down time of the processor, if
+        set."""
+        return self._expected_down_time
+
+    def expected_recovery_time(self) -> 'Optional[datetime.datetime]':
+        """Returns the expected the processor should be available, if set."""
+        return self._expected_recovery_time
+
+    def _create_id(self, id_type: str = 'reservation') -> str:
```

Comment:
It sounds like this will work from your comment, but to confirm: will the following behave properly?
```
id1 = self._create_id()                          # id1 = '.../1/reservation'
id2 = self._create_id(id_type='something_else')  # id2 = '.../2/something_else'
id3 = self._create_id()                          # id3 = '.../3/reservation'
```
i.e. is it OK that the two 'reservation' IDs don't have consecutive numbers?

Reply:
Yeah, this should be fine.  I think the reality is that the ids are UUIDs, not consecutive integers.  We just want them to be unique.  I think the one issue is that jobs should really be children of programs, but I am not going to worry about that right now.